File: C#/DataStructures/DataStructures/sorts/SelectionSort.cs
using System;

namespace DataStructures.sorts
{
    public static class SelectionSort
    {
        public static T[] Sort<T>(T[] arr)
            where T : IComparable
        {
            T[] output = new T[arr.Length];
            arr.CopyTo(output, 0);

            for(int i = 0; i < output.Length - 1; i++)
            {
                int min = i;

                for(int j = i; j < output.Length; j++)
                {
                    if (output[j].CompareTo(output[min]) < 0)
                        min = j;
                }

                var tmp = output[i];
                output[i] = output[min];
                output[min] = tmp;
            }

            return output;
        }
    }
}

File: Scala/src/main/com/ckeyser/sorts/SelectionSort.scala
package com.ckeyser.sorts

import com.ckeyser.util.SortUtil

import Ordering.Implicits._

/**
  * Created by Chris Keyser on 10/2/2016.
  */
object SelectionSort {
    def sort[A : Ordering](arr: Array[A]): Array[A] = {
        val out = arr.clone()

        for(i <- out.indices) {
            var min = Int.MaxValue
            // find the next lowest element
            for(j <- Range(i, out.length)) {
                min = if (out(j) < out(min)) j else min
            }

            SortUtil.swap(out, min, i)
        }

        out
    }
}
